cpp/src/arrow/compute/kernels/vector_cumulative_ops_test.cc

```diff
@@ -0,0 +1,257 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include <gtest/gtest.h>
+
+#include "arrow/array.h"
+#include "arrow/chunked_array.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/testing/util.h"
+#include "arrow/type.h"
+
+#include "arrow/compute/api.h"
+#include "arrow/compute/kernels/test_util.h"
+
+namespace arrow {
+namespace compute {
+
+template <typename T, typename OptionsType>
+class TestCumulativeOp : public ::testing::Test {
+ public:
+  using ArrowType = T;
+  using ScalarType = typename TypeTraits<T>::ScalarType;
+  using CType = typename TypeTraits<T>::CType;
+
+ protected:
+  std::shared_ptr<DataType> type_singleton() { return default_type_instance<T>(); }
+
+  std::shared_ptr<Array> array(const std::string& values) {
+    return ArrayFromJSON(type_singleton(), values);
+  }
+
+  std::shared_ptr<ChunkedArray> chunked_array(const std::vector<std::string>& values) {
+    return ChunkedArrayFromJSON(type_singleton(), values);
+  }
+
+  template <typename V = T>
+  enable_if_t<!is_floating_type<V>::value, void> Assert(
+      const std::string func, const std::shared_ptr<Array>& input,
+      const std::shared_ptr<Array>& expected, const OptionsType& options) {
+    ASSERT_OK_AND_ASSIGN(auto result, CallFunction(func, {Datum(input)}, &options));
+
+    AssertArraysEqual(*expected, *result.make_array(), false, EqualOptions::Defaults());
+  }
+
+  template <typename V = T>
+  enable_if_t<!is_floating_type<V>::value, void> Assert(
+      const std::string func, const std::shared_ptr<ChunkedArray>& input,
+      const std::shared_ptr<ChunkedArray>& expected, const OptionsType& options) {
+    ASSERT_OK_AND_ASSIGN(auto result,
+                         CallFunction(func, {Datum(input)}, &options, nullptr));
+
+    ChunkedArray actual(result.chunks(), this->type_singleton());
+    AssertChunkedEqual(*expected, actual);
+  }
+
+  template <typename V = T>
+  enable_if_floating_point<V> Assert(const std::string func,
```

Comment:
You shouldn't have to special-case floating point, just use `AssertArraysApproxEqual` always.